### PR TITLE
Add support to show suggestions on CommandNotFound exception

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
## PR Summary

Add `-UseFuzzyMatching` switch to `Get-Command` to return commands that fuzzy match to the name in order of closest match to least.  Integrate this with the suggestion framework so that on a `CommandNotFound` exception, suggestions are shown for likely commands the user intended.  The suggestions are part of the console host and write directly to the console so don't see a way to test that.  In the future, we should consider changing that to using information stream.

Example usage

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing https://github.com/PowerShell/PowerShell-Docs/pull/3429
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by iSazonov** [comment]
We could consider other names of the parameter: -Fuzzy, -FuzzyMatching, -ShowSuggestions, -Suggestions.

We need PowerShell Committee approvement for the new public API.

**Comment by rjmholt** [comment]
We could introduce a test hook to do something like write output to the information stream or to a variable for testing purposes for now

File: src/System.Management.Automation/utils/FuzzyMatch.cs
Comment (PaulHigin) [inline src/System.Management.Automation/utils/FuzzyMatch.cs]: Is this settable by the user?

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/GetCommandCommand.cs
Comment (rjmholt) [reply src/System.Management.Automation/engine/GetCommandCommand.cs]: C# [has named tuples](https://docs.microsoft.com/en-us/dotnet/csharp/tuples#named-and-unnamed-tuples), but only for value tuples.

I imagine it's possible to add these as value tuples to the data structure, but by then I think a private class maybe is about the same amount of work?

File: src/System.Management.Automation/engine/hostifaces/HostUtilities.cs
Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/hostifaces/HostUtilities.cs]: It's a bit inconsistent.  I believe CodeFactor (aka StyleCop) doesn't like the `s_Name` prefix.

File: src/System.Management.Automation/utils/FuzzyMatch.cs
Comment (SteveL-MSFT) [reply src/System.Management.Automation/utils/FuzzyMatch.cs]: No.  It's an implementation detail and if we ever change the algorithm, this setting may not even make sense.

File: src/System.Management.Automation/engine/GetCommandCommand.cs
Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/GetCommandCommand.cs]: I believe I had `var` first, but compiler complained.

**Comment by SteveL-MSFT** [comment]
@rjmholt I suggest we defer the test hook as that would make sense for testing suggestions in general and not this specific change.

**Review by PaulHigin — APPROVED** [review]

**Comment by iSazonov** [comment]
It was good to have a tracking issue if we needed these special tests.

File: src/System.Management.Automation/engine/hostifaces/HostUtilities.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/hostifaces/HostUtilities.cs]: Our code guideline https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/coding-guidelines.md#naming-conventions requires this style. We could ignore CodeFactor warnings until we align its settings.

Comment (rjmholt) [reply src/System.Management.Automation/engine/hostifaces/HostUtilities.cs]: I like `s_staticField`. I find it very helpful when reasoning about the lifetime of the object. I've can't recall anywhere we don't use it, although I haven't been touching PowerShell so much recently.

I know it gets labelled as "Hungarian notation" and dismissed as unneeded in the world of IDEs. But (1) I am confronted by a lack of IDE support all too often and (2) Hungarian notation is historically for type information, which is not helpful in OOP -- instead I expect the variable to be named meaningfully -- whereas scope is still something I have to think about orthogonally while writing C#.

My preference would be for CodeFactor to promote its usage.

File: src/System.Management.Automation/engine/GetCommandCommand.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/GetCommandCommand.cs]:
```suggestion
        private List<(CommandInfo command, int matchScore)> _commandScores = new List<(CommandInfo command, int matchScore)>();
```

File: src/System.Management.Automation/engine/GetCommandCommand.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/GetCommandCommand.cs]:
```suggestion
                results = _commandScores.OrderBy(x => x.matchScore).Select(x => x.command).ToList();
```

File: src/System.Management.Automation/engine/GetCommandCommand.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/GetCommandCommand.cs]: Given that the lexical stack is now `method` > `foreach` > `try` > `if` > `else if` > `if` > `if`, I think we should seriously consider spinning out at least one method here. One that returns an `IEnumerable<CommandInfo>` maybe?

File: src/System.Management.Automation/engine/GetCommandCommand.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/GetCommandCommand.cs]:
```suggestion
                                    commands = _commandScores.Select(x => x.matchScore).ToList();
```

File: src/System.Management.Automation/engine/GetCommandCommand.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/GetCommandCommand.cs]:
```suggestion
                            _commandScores.Add((command: current, matchScore: score));
```

File: src/System.Management.Automation/engine/Modules/ModuleUtils.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/Modules/ModuleUtils.cs]:
```suggestion
                    yield return (command: command, matchScore: score);
```

File: src/System.Management.Automation/engine/hostifaces/HostUtilities.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/hostifaces/HostUtilities.cs]: I think the newline here makes the enum less cohesive to read. My feeling is that the values should either all have newlines (and comments) or all be together.

File: src/System.Management.Automation/engine/hostifaces/HostUtilities.cs
Comment (rjmholt) [reply src/System.Management.Automation/engine/hostifaces/HostUtilities.cs]: I like `s_staticField`. I find it very useful when reading code to reason about the scope and lifetime of a field. I also can't recall any static fields in PowerShell that don't start with `s_`.

I know "Hungarian notation" is frowned upon as unmodern now, but:
1. I don't think of scope-prefixing as HN -- in OOP classes are types, so giving the variable a good, semantic name is better than HN -- but I still have to think about scope and lifetime orthogonally to the type of the object, so the `_` and `s_` are very useful.
1. IDEs only outmode HN if I have an IDE and it works. A lot of the time I spend reading code is on GitHub, reviewing or suggesting edits, and even when I'm in an IDE, type inference and completions are just not all that reliable.

Ultimately though, I think if `_` is required then `s_` should be too, any criticism applying to the latter also applies to the former.

So I think CodeFactor should require `s_`.

**Comment by SteveL-MSFT** [comment]
@iSazonov created https://github.com/PowerShell/PowerShell/issues/8486

File: src/System.Management.Automation/utils/FuzzyMatch.cs
Comment (rjmholt) [inline src/System.Management.Automation/utils/FuzzyMatch.cs]: Worth adding info as to where this implementation comes from

File: src/System.Management.Automation/engine/hostifaces/HostUtilities.cs
Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/hostifaces/HostUtilities.cs]: Will update

File: src/System.Management.Automation/engine/hostifaces/HostUtilities.cs
Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/hostifaces/HostUtilities.cs]: Will add comments to each value

File: src/System.Management.Automation/utils/FuzzyMatch.cs
Comment (SteveL-MSFT) [reply src/System.Management.Automation/utils/FuzzyMatch.cs]: It's in the comment at the top of the file.  I'll move it to the summary here.

**Comment by TravisEz13** [comment]
@SteveL-MSFT 	Do you want to wait on Rob to review your changes?

**Comment by SteveL-MSFT** [comment]
@TravisEz13 if @rjmholt has further comments, we can treat it as a separate issue and PR

**Comment by iSazonov** [comment]
It would be good to fix some style issues and "s_" before merge.

File: src/System.Management.Automation/engine/Modules/ModuleUtils.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/Modules/ModuleUtils.cs]:
```suggestion
            Score =  score;
```

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this and is ok with `-UseFuzzyMatching`

**Review by rjmholt — APPROVED** [review]

**Pull request merged**
